iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp

```diff
@@ -0,0 +1,228 @@
+// Copyright (c) 2020 by Robert Bosch GmbH. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iceoryx_utils/cxx/function_ref.hpp"
+#include "test.hpp"
+
+using namespace ::testing;
+using namespace iox::cxx;
+
+int freeFunction()
+{
+    return 42 + 42;
+}
+
+class Functor
+{
+  public:
+    int operator()()
+    {
+        return 73;
+    }
+};
+
+struct ComplexType
+{
+    char a;
+    int b;
+    float c;
+
+    bool operator==(const ComplexType& rhs) const
+    {
+        if (a == rhs.a && b == rhs.b && c == rhs.c)
+        {
+            return true;
+        }
+        else
+        {
+            return false;
+        }
+    }
+};
+
+ComplexType returnComplexType(ComplexType foo)
+{
+    return foo;
+}
+
+class function_refTest : public Test
+{
+  public:
+    void SetUp() override
+    {
+    }
+
+    void TearDown() override
+    {
+    }
+
+    int foobar()
+    {
+        return 4273;
+    }
+
+    uint8_t m_iter{0};
+};
+
+using function_refDeathTest = function_refTest;
+
+TEST_F(function_refTest, CreateEmpty)
+{
+    function_ref<void()> sut;
+    EXPECT_FALSE(sut);
+}
+
+TEST_F(function_refTest, CreateWithNullptr)
+{
+    function_ref<void()> sut(nullptr);
+    EXPECT_FALSE(sut);
+}
+
+TEST_F(function_refDeathTest, CreateEmptyLeadsToTermination)
+{
+    function_ref<void()> sut;
+    EXPECT_DEATH(sut(), ".*");
+}
+
+TEST_F(function_refTest, CreateEmptyAndAssign)
+{
+    auto lambda = [] {};
+    function_ref<void()> sut;
+    sut = lambda;
+    EXPECT_TRUE(sut);
+}
+
+TEST_F(function_refTest, CreateAndCopy)
+{
+    auto lambda = []() -> int { return 42; };
+    function_ref<int()> sut1(lambda);
+    function_ref<int()> sut2{sut1};
+    EXPECT_TRUE(sut2);
+    EXPECT_THAT(sut2(), Eq(42));
+}
+
+TEST_F(function_refTest, CreateAndCopyAssign)
```

Comment:
Maybe you should also add a tests where you copy a function_ref into an invalid/empty function or where you copy an invalid function_ref into a valid one.

Reply:
Added